### PR TITLE
SPS: Allow configuring whether ring sockets are bidirectional

### PR DESCRIPTION
**Pull request by TheLastRar**
While bidirectional ring sockets are useful for hands, rings sockets are also used where bidirectional support is either unneeded or unwanted.
Additionally, supporting bidirectional appears to require stricter limits on how the plug tracks the socket, vs holes

Use a slightly offset the range set to indicate if a socket is bidirectional, while still retaining compatibility with DPS & TPS

~~Additionally, we change the equation for light range to a supposedly more accurate one as per this [post from a Unity dev](https://forum.unity.com/threads/point-light-in-v-f-shader.499717/#post-9052987)~~ (Already in main)

Note that this makes ring orifices non reversible by default.

~~TODO Decide what we want to do on upgrade,
Do we want to change all rings to bidirectional? or only those on the hands?~~
All existing VRCFuryHapticSocket Rings are set as bidirectional, (as VRCFArmatureUtils uses editor only APIs)

```
The changes made in this contribution are free and unencumbered software
released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <https://unlicense.org>
```

### CODE REVIEW

File: com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryHapticSocketEditor.cs [hunk 1]
```diff
@@ -277,7 +290,8 @@ public class VRCFuryHapticPlaySocketEditor : UnityEditor.Editor {
                     var mainLight = main.AddComponent<Light>();
                     mainLight.type = LightType.Point;
                     mainLight.color = Color.black;
-                    mainLight.range = lightType == VRCFuryHapticSocket.AddLight.Ring ? 0.4202f : 0.4102f;
+                    mainLight.range = lightType == VRCFuryHapticSocket.AddLight.Ring ? 0.4202f :
+                        lightType == VRCFuryHapticSocket.AddLight.RingBidirectional ? 0.4201f : 0.4102f;
```
Comment (TheLastRar) [inline com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryHapticSocketEditor.cs:294]: I'm unsure as to how the SPSLL subtype and the bidirectional subtype should interact, maybe treat them as flags?

File: com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryHapticSocketEditor.cs [hunk 2]
```diff
@@ -422,6 +438,12 @@ public class VRCFuryHapticPlaySocketEditor : UnityEditor.Editor {
             return ShouldProbablyHaveTouchZone(socket);
         }
 
+        public static bool ShouldProbablyBeReversible(VRCFuryHapticSocket socket) {
+            if (HapticUtils.IsChildOfBone(socket.owner(), HumanBodyBones.LeftHand)) return true;
```
Comment (SenkyDragon) [inline com.vrcfury.vrcfury/Editor/VF/Inspector/VRCFuryHapticSocketEditor.cs:442]: I don't think we should do a guess here. The vast majority of rings should be double sided, including ones that don't wind up on hands. I don't think it's unreasonable to make people select single sided if they choose for some specific use case.

**Comment by SenkyDragon** [comment]
Mostly lgtm, but I think the default behavior should be bidirectional. Uni-directional is more niche, and bidirectional has been the default in SPS since the beginning.

**Comment by TheLastRar** [comment]
> Mostly lgtm, but I think the default behavior should be bidirectional. Uni-directional is more niche, and bidirectional has been the default in SPS since the beginning.

I've made Auto and all upgrade rings bidirectional, rather than guessing.

Are you me to invert the bidirectional flag (for light tracking), making it a uni-directional flag?
I'm hesitant to do that as even outside of the niche SPS prefab as I've run into issues with bidirectional rings flipping direction.
DPS and TPS avatars would not be able to opt in/out and would be stuck with whatever a SPS plug decides to do by default.
Although, I could have it apply to SPSLL tagged sockets by default (turning the SPSLL flag into a supports SPS extensions flag), with the directional flag making it unidirectional. if that is preferable.

**Comment by SenkyDragon** [comment]
Thank you for your contribution and patience. I did wind up rewriting quite a lot of this to work in the new SPS+ "semi-lightless" system, which means that we actually don't change the light range at all. For now, this does require the user to opt into the `SPS+` toggle in their in-game SPS Options menu, but once SPS+ is more vetted, it will become the default and the option will go away.

**Comment by TheLastRar** [comment]
Thanks, I hadn't had the chance to go over this (or do any a avatar work) during last week when SPS+ rolled out
I see legacy rings are still forced as two-way, maybe I had taken the wrong approach for this.

Nice that you where able to avoid the extra contact receiver on the plug.

I did, however, run into some issues when testing https://github.com/VRCFury/VRCFury/issues/185

**Pull request merged**